File: clojure/src/advent2016/day6.clj
(ns advent2016.day6
  (:require [advent.core :refer [->lines]]
            [advent2016.core :refer [read-puzzle]]))

(def example1 ["eedadn" "drvtee" "eandsr" "raavrd" "atevrs" "tsrnev" "sdttsa"
               "rasrtv" "nssdts" "ntnada" "svetve" "tesnvt" "vntsnd" "vrdear"
               "dvrsen" "enarar"])
(def puzzle (read-puzzle "day6" ->lines))
#_(def puzzle (vec (line-seq (clojure.java.io/reader "../day6.data"))))

(def solution1 (apply str (map #(->> (frequencies %)
                                     (sort-by (comp - val))
                                     ffirst)
                               (apply map str puzzle))))

(def solution2 (apply str (map #(->> (frequencies %)
                                     (sort-by val)
                                     ffirst)
                               (apply map str puzzle))))

File: python/day6.py
import sys
from functools import partial
from util import frequencies

example1 = ["eedadn", "drvtee", "eandsr", "raavrd", "atevrs", "tsrnev",
            "sdttsa", "rasrtv", "nssdts", "ntnada", "svetve", "tesnvt",
            "vntsnd", "vrdear", "dvrsen", "enarar"]


def ffreq(freqs, reverse):
    return sorted(freqs.items(), key=lambda (k, v): v, reverse=reverse)[0][0]


def solution1(input):
    return ''.join(
        map(partial(ffreq, reverse=True), map(frequencies, zip(*input))))


def solution2(input):
    return ''.join(
        map(partial(ffreq, reverse=False), map(frequencies, zip(*input))))


if __name__ == '__main__':
    input = list(sys.stdin)
    print solution1(input)
    print solution2(input)
